### PR TITLE
Fixes LRP service deletion/restoration and adds e2e tests for LRP.

### PR DESCRIPTION
**Pull request by Weil0ng**
Signed-off-by: Weilong Cui <cuiwl@google.com>

### CODE REVIEW

**Comment by joestringer** [comment]
test-focus K8s.*LRP

File: test/k8sT/manifests/lrp-test.yaml
Comment (aanm) [inline test/k8sT/manifests/lrp-test.yaml]: leftover?

**Review by aanm — CHANGES_REQUESTED** [review]

File: test/k8sT/Services.go
Comment (brb) [inline test/k8sT/Services.go]: LRP depends on the host-reachable services feature which is enabled only on `net-next` and `4.19` CI runs. To skip the suite on other CI runs of other types, you can use `SkipContextIf(func () bool { return !helpers.RunsOnNetNextOr419Kernel() }, "Checks local redirect policy", func () { ...`.

File: test/k8sT/Services.go
Comment (brb) [inline test/k8sT/Services.go]: Nit: I think this can be declared as a const before `BeforeAll(..)`, can't it?

File: test/k8sT/Services.go
Comment (brb) [inline test/k8sT/Services.go]: Maybe then it makes sense to add a failing assertion instead of returning the empty string?

File: test/k8sT/Services.go
Comment (brb) [inline test/k8sT/Services.go]: Shall we grep for `clusterIP` to make sure that that service is of the LR type?

File: test/k8sT/Services.go
Comment (brb) [inline test/k8sT/Services.go]: Alternatively, you could check against pod name. Echoserver (both - tcp and udp) should return `Hostname: ...` which contains a pod name.

**Review by brb — COMMENTED** [review]
Thanks!

File: test/k8sT/Services.go
Comment (aditighag) [inline test/k8sT/Services.go]: Just to confirm this is only meant to simply the test workflow, right? LRP manager does process multiple podIPs.

File: test/k8sT/manifests/lrp-test.yaml
Comment (aditighag) [inline test/k8sT/manifests/lrp-test.yaml:5]: The port names don't match with the ones specified in the yaml, is that intentional? Maybe this is the remote be? On a related note, does it make sense to name the pods as local-backend and remote-backend?

File: test/k8sT/manifests/lrp-test.yaml
Comment (Weil0ng) [reply test/k8sT/manifests/lrp-test.yaml]: Done.

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go]: Ah makes sense, thanks!

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go]: Done.

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go]: I think the assertion is covered before the loop above, this is just to make compiler happy, but maybe it's a bad practice?

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go]: This is after LRP has been deployed, hence the service type should be of string "LocalRedirect", are you suggesting we check service type *before* deploying LRP too?

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go]: Yup, this is just to make testing a bit simpler, we can probably add a test case with multiple backends in follow up PR?

File: test/k8sT/manifests/lrp-test.yaml
Comment (Weil0ng) [reply test/k8sT/manifests/lrp-test.yaml:5]: > The port names don't match with the ones specified in the yaml

I'm a bit confused, which port name are you referring to?
 
> does it make sense to name the pods as local-backend and remote-backend

Since "local" and "remote" are relative to the requester, I think maybe we should keep the name generic here?

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go]: Thanks! Changed to check the returned pod name, I think that's better :)

File: test/k8sT/Services.go
Comment (joestringer) [inline test/k8sT/Services.go:423]: The [GSG](https://docs.cilium.io/en/latest/gettingstarted/local-redirect-policy/) doesn't mention anything about kernel version, neither does the [Kernel version requirements](https://docs.cilium.io/en/latest/operations/system_requirements/#required-kernel-versions-for-advanced-features). Is there a reason we are restricting the versions we run this test on?

File: test/k8sT/Services.go
Comment (joestringer) [inline test/k8sT/Services.go]: The `cancel` function must be kept & called to clean up context resources.

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go:423]: Quoting @brb "LRP depends on the host-reachable services feature which is enabled only on net-next and 4.19 CI runs."

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go]: Done.

File: test/k8sT/Services.go
Comment (joestringer) [reply test/k8sT/Services.go:423]: Ok, then might be worth following up to make this clearer in the docs.

Comment (aditighag) [reply test/k8sT/Services.go:423]: That's right. I've already updated the documents in the follow-up PR. For the socket lookup call, we need at least kernel version 5.1, but I checked the other day, and the default in GKE seems to be 4.19, which is okay for now.

File: test/k8sT/Services.go
Comment (brb) [reply test/k8sT/Services.go]: I meant to grep for a particular service to see whether it's type has changed to LocalRedirect, i.e. `fmt.Sprintf("cilium service list | grep %s", clusterIP)`. Because right now we are checking whether any service is of the LocalRedirect type.

File: test/k8sT/Services.go
Comment (aditighag) [reply test/k8sT/Services.go:423]: It's worth finding out if we need the host-reachable functionality for the node-local dns case. @Weil0ng Can you revert the kernel check, and trigger CI test suite with `test-me-please`/`retest-4.9`?

Comment (aditighag) [reply test/k8sT/Services.go:423]: @brb confirmed that the host-reachable functionality is needed for being able to do service translation from host ns. This would apply for pods deployed in the `hostNetwork` mode. So the minimum kernel version required would indeed be 4.19.

Comment (brb) [reply test/k8sT/Services.go:423]: > So the minimum kernel version required would indeed be 4.19.

To be more precise - 4.19.57 for UDP, and 4.17.0 for TCP.

Comment (aditighag) [reply test/k8sT/Services.go:423]: Yes, I've kept the version requirement same as kubeproxy-free/host-services features - https://github.com/cilium/cilium/pull/13362/commits/9289cb65f2b0668addb7a9a19758c42cd4419e93

Comment (Weil0ng) [reply test/k8sT/Services.go:423]: > It's worth finding out if we need the host-reachable functionality for the node-local dns case.

As @aditighag and @brb pointed out, I think for node-local dns, we need host-reachable services too to capture dns requests from host-ns pods.

**Comment by joestringer** [comment]
test-focus K8s.*LRP

**Review by aanm — APPROVED** [review]

**Review by brb — CHANGES_REQUESTED** [review]
Some comments.

**Comment by Weil0ng** [comment]
test-focus K8s.*LRP

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go]: Ah I see, done.

**Comment by Weil0ng** [comment]
Addressed comments, added a new test case to test for service restoration (PR #13274), PTAL :)

**Review by aditighag — APPROVED** [review]

File: test/k8sT/manifests/lrp-test.yaml
Comment (aditighag) [inline test/k8sT/manifests/lrp-test.yaml]: This won't work, and we need to check if/how CI test is passing. When we have multiple ports, the ports need to be named, and they need to match the ones specified in the pod spec.  The port names are then used to match frontend ports to the corresponding backend ports. 
This is documented in gsg, CRD spec in addition to the code.
@Weil0ng Were you able to run this test successfully?

**Review by aditighag — CHANGES_REQUESTED** [review]
Check last comment. Please don't merge until we sort out the issue/get a confirmation that the test is set up correctly.

File: test/k8sT/manifests/lrp-test.yaml
Comment (Weil0ng) [reply test/k8sT/manifests/lrp-test.yaml:5]: Great catch, done :)

File: test/k8sT/manifests/lrp-test.yaml
Comment (Weil0ng) [reply test/k8sT/manifests/lrp-test.yaml]: Fixed, I can run the test steps manually, but my local e2e test setup is broken (not due to this PR)...

**Comment by aditighag** [comment]
test-focus K8sServicesTest

File: test/k8sT/manifests/lrp-test.yaml
Comment (aditighag) [reply test/k8sT/manifests/lrp-test.yaml]: Let's get a confirmation from the automated CI run.

File: test/k8sT/Services.go
Comment (brb) [inline test/k8sT/Services.go]: Nit: `for try := 0; try < repeat; try++ { ... }`

File: test/k8sT/Services.go
Comment (brb) [inline test/k8sT/Services.go]: Nit: small comment saying that you are expecting to see pod name in the returned Hostname field would be useful.

**Review by brb — APPROVED** [review]
Two nits. Feel free to address them in a follow up PR.

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go]: Done.

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go]: Done.

**Comment by Weil0ng** [comment]
test-focus K8sServicesTest

**Comment by Weil0ng** [comment]
test-focus K8sServicesTest

**Comment by Weil0ng** [comment]
test-focus K8s.*LRP

**Comment by Weil0ng** [comment]
test-focus K8s.*LRP

**Comment by Weil0ng** [comment]
test-focus K8s.*LRP

File: test/k8sT/Services.go
Comment (christarazi) [inline test/k8sT/Services.go]: Nit:

```suggestion
			for _, pod := range []string{feFilter, beFilter} {
```

File: test/k8sT/Services.go
Comment (christarazi) [inline test/k8sT/Services.go]: I don't think this needs to be an `ExpectWithOffset`. That's reserved for when you have a helper function that's called from inside an `It()` block which performs an assertion. Hence, needing an offset because it's inside a function, rather than in the `It()` itself. This is relevant when the test fails and you get a stacktrace of where the failure occurred.

**Comment by Weil0ng** [comment]
test-focus K8s.*LRP

File: test/k8sT/Services.go
Comment (christarazi) [inline test/k8sT/Services.go]: Maybe we don't need to do a boolean AND here. Probably an individual assertion on both should be sufficient, and removes the need to pretty print the error msg.

File: test/k8sT/Services.go
Comment (christarazi) [inline test/k8sT/Services.go]: Why do we need to do retries? Is this something that will eventually succeed? If so, then please use `Eventually` for this.

**Review by christarazi — CHANGES_REQUESTED** [review]
A few comments from me regarding testing logic and suggestions to improve / simplify the tests.

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go]: Done.

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go]: Done.

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go]: Good point, done :)

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go]: hmm, not sure if `Eventually` is suitable here. This is to check that lb will "always" send traffic to local backend and "never" send it to remote backend.

Comment (christarazi) [reply test/k8sT/Services.go]: Ah, I see, so then it's the opposite! We can use `Consistently` to do this :wink:

Comment (Weil0ng) [reply test/k8sT/Services.go]: makes sense, I think consistently would work for this test, but I doubt either `consistently` or `eventually` matches the semantics in the restore test (line 533), where we need results from multiple curl results to "sometimes" have substring1 and "sometimes" have substring2, any magic work/pattern to use there?

Comment (Weil0ng) [reply test/k8sT/Services.go]: ah is that just  an `Eventually` && `Eventually` case?

**Comment by Weil0ng** [comment]
test-focus K8s.*LRP

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go]: Done.

**Comment by Weil0ng** [comment]
test-focus K8s.*LRP

**Comment by Weil0ng** [comment]
test-focus K8s.*LRP

File: test/k8sT/Services.go
Comment (christarazi) [inline test/k8sT/Services.go]: I think you meant this?
```suggestion
					return strings.Contains(returned, tc.want) && !strings.Contains(returned, tc.notWant)
```

**Review by christarazi — APPROVED** [review]
LGTM! The use of `Consistently` and `Eventually` simplify the tests nicely. Just one potential typo above.

File: test/k8sT/Services.go
Comment (Weil0ng) [reply test/k8sT/Services.go]: urghhh, fat fingers :) Fixed.

**Comment by Weil0ng** [comment]
test-focus K8s.*LRP

**Comment by Weil0ng** [comment]
Updated pod execution context deadline from 1s to 30s to reduce flakiness.

**Comment by Weil0ng** [comment]
test-focus K8s.*LRP

**Comment by Weil0ng** [comment]
test-focus K8s.*LRP

**Comment by Weil0ng** [comment]
test-focus K8s.*LRP

**Comment by Weil0ng** [comment]
test-focus K8s.*LRP

**Comment by Weil0ng** [comment]
test-focus K8s.*LRP

**Comment by Weil0ng** [comment]
test-focus K8s.*LRP

**Comment by Weil0ng** [comment]
test-focus K8s.*LRP

**Comment by Weil0ng** [comment]
test-me-please

**Review by aditighag — APPROVED** [review]
LGTM. Please add unit tests as a follow-up for the restoration logic to catch the issues that we were debugging yesterday, in case the code path is not covered by the e2e test. For example, assert that service restoration happens only once per policy config when it's deleted.

**Comment by aditighag** [comment]
CI test failure is legit - LRP test is still failing.

**Comment by aditighag** [comment]
Don't merge the PR until the LRP CI test is fixed.

**Comment by Weil0ng** [comment]
Needs to rerun CI after PR #13494 is merged.

**Comment by Weil0ng** [comment]
test-me-please

**Review by aditighag — APPROVED** [review]
We can merge the PR once the LRP test passes.

**Comment by Weil0ng** [comment]
The added LRP tests are passing with 4.19 kernels, the two failing CI seem unrelated to the change.

**Pull request merged**
